### DISCUSSION TITLE
Is there way to view headings using telescope ?

### DISCUSSION THREAD

**Discussion by aniketgm** [nvim-telekasten/telekasten.nvim#210] [Q&A]
As mentioned in the title, Is there a way to view the headings of a file using telescope. I am aware about the [Link to specific heading](https://github.com/renerocksai/telekasten.nvim/discussions/14) feature. However, the use-case I have in mind is somewhat like,
* View the headings in telescope
* Search for a specific heading and jump to it.

So I was wondering if this is possible in telekasten ?

**Comment by lambtho12** [comment]
Let me know if I get this right. The process you have in mind is something like:

1. Use `find_note` or `grep_note` or something like that to find a specific note
2. Instead of pressing enter and open the note, press a key and have a new telescope picker that would extract the headings of that specific note and let you browse it
3. Press enter to open this file and go the the heading, or have the other actions for the link to the heading specifically (ctrl+y to yank and ctrl+i to insert). (Cherry on the cake would be also being able to press escape and close the header picker to return to the find note picker).


This is currently not possible. But it should be doable (and very useful). The absolute best would be to have the picker ordering the entries by order of appearance in the text, so it would act similarly to a table of contents and help you find the proper heading even faster.

**Reply by aniketgm** [reply]
@lambtho12 What you described would be even more nice !! Cherry on top 👍 
The use-case I was thinking was for when we are inside just one markdown file (I mean the active buffer), where we search the headings only within that file and jump to it.

Suggestions:
1. Use-case 1: Let's say I want to see only headings within the current markdown file buffer,
    * Then, similar to what we currently have `:Telekasten show_tags`, we could have something like `:Telekasten show_headings` for the active md file.
2. Use-case 2: I want to jump to a different file (`:Telekasten find_notes`) and also to a different heading, whilst finding the file,
    * Then, @lambtho12's suggestion mentioned above.

**Reply by lambtho12** [reply]
First use case can already be accomplished by markdown-specific plugins in general. But I reckon it would be a nice addition as it would not require a user to install an entire plugin just for this functionality. Besides, it is more or less a prerequisite in order to implement use-case 2.

Both use cases seem good to me. Could you open two separate FR for that?

I will try to set a full day next week to work on telekasten. I should be able to cram that in.

**Reply by aniketgm** [reply]
FR #213 created. two FRs might not be required.
Not sure of how lua works, if its possible then I think the solution code from the "View headings from same file" could be used for the use-case no.2. So both scenarios would be covered.

**Reply by lambtho12** [reply]
> the solution code from the "View headings from same file" could be used for the use-case no.2. 

Yep, this is definitely how it is going to be used. A function like `show_heading(file)` that throws a picker for the file can be used for any file. So we can reuse this for both senarii

**Comment by aleksey-rowan** [comment]
There is a Telescope plugin to just between Markdown headings: https://github.com/crispgm/telescope-heading.nvim Would this help?

**Reply by aniketgm** [reply]
> There is a Telescope plugin to just between Markdown headings: https://github.com/crispgm/telescope-heading.nvim Would this help?

Thanks a lot for this. 💯 This surely helps.. 

Although, this will add some unnecessary extra code like for e.g. it provides support for neorg, beancount, etc, which I will never really use, But (for now) it is fine for me.
Also, as mentioned in above discussion, I think, the use-case no. 2:
> Use-case 2: I want to jump to a different file (:Telekasten find_notes) and also to a different heading, whilst finding the file,
Then, @lambtho12's suggestion mentioned above.

This is specific to telekasten and would be a really cool feature to have.

**Reply by lambtho12** [reply]
Awesome! Thank you very much. 
As @aniketgm points out, it only partially achieves what we want here. But I will definitely ~steal~ look for inspiration in this function.

**Comment by cnshsliu** [comment]
I am writing some codes to implement:

1. Note structure (tags, links, todos) and hierarchy (parent and children) navigation
2. Search notes incrementally, by either tag or content
3. Search an AND or OR query (match all words or any of them)
4. Exclude words (match those, but not these)
5. Saved search history, easily pick up again.
6. Headers outline and jumping

See preview screenshots:

<img width="1722" alt="image" src="https://user-images.githubusercontent.com/2124836/224567639-40b2244b-1156-4042-931f-0dd74b5db71b.png">
<img width="1715" alt="image" src="https://user-images.githubusercontent.com/2124836/224567699-b88c56cc-9dfb-4eed-8f18-0468ebfbbfd3.png">
<img width="1711" alt="image" src="https://user-images.githubusercontent.com/2124836/224568371-3353f170-522a-4b69-8e5c-1cf36e02749a.png">
